src/directLine.ts

```diff
@@ -214,4 +226,92 @@ export class DirectLine implements IBotConnection {
+                        timeout,
+                        headers: {
+                            "Accept": "application/json",
+                            "Authorization": `Bearer ${this.token}`
+                        }
+                    })// Takes one result
+                    .take(1)
+                    // Extracts the stream URL
+                    .map(result => result.response.streamUrl);
+                }
+            })            
+            // Retries every Timeout seconds in case of connection issue other than security
+            .retryWhen(error$ => error$
+                .mergeMap(error => {
+                    if (error.status === 403) {
+                        this.connectionStatus$.next(ConnectionStatus.Offline);
+                        return Observable.throw(error);
+                    } else {
+                        return Observable.of(error);
+                    }
+                })
+                .delay(timeout)
+            );
+    }
+
+    /**
+     * Gets an Observable of activity from the Web Socket connectivity of direct line.
+     * @return An Observable of activity.
+     */
+    private getWebSocketActivity$(): Observable<Activity> {
+        // From the Web Socket stream URL
+        return this.getWebSocketStreamURL$()
+            // Create an observable Web Socket
+            .flatMap<string, WebSocketResponse>(url => {
+                // Use a result selector to prevent crash from incoming empty message (it could also helps in timeout detection)
+                var ws$ = Observable.webSocket({
+                    url: url,
+                    resultSelector: e => ({                 
+                        type: e.data ? "ACTIVITY" : "PING",
+                        message: e.data ? JSON.parse(e.data) : null
+                    })
+                });
+
+                // timeout seconds ping for keep alive even if already in the browser.
+                Observable.interval(timeout)
+                    .timeInterval()
+                    .subscribe(_ => { ws$.next({}) });
+
+                return ws$;                
+            })
+            // Retry the full connection + grabbing the socket URL in case of issue (reset the connection) 
+            .retryWhen(error$ => error$
+                .mergeMap(error => {
+                    return Observable.of(error);
+                }) 
+                .delay(timeout))
+            // Only deals with activity message (discard the ping)
+            .filter(data => data.type == "ACTIVITY")
+            // Maps to activity group
+            .map(data => data.message)
+            // Creates observable of activityfor the client as well as updating the watermark           
+            .flatMap(activityGroup => {
+                if (activityGroup.watermark)
+                    this.watermark = activityGroup.watermark;
+                return Observable.from(activityGroup.activities);
+            });
```

Comment:
This is probably not necessary, but what do you think of extracting this little bit of code, which is shared with the getPolling code, into its own function? We'd end up with something like `flatMap(activityGroup => observableFromActivityGroup(activityGroup))` - a little less repetition, a little more expressive. I'm cool either way.